# Denovo_assembly_flye

## Data Analysis
Programs required: it is recommended that the user has anaconda installed, through which all required programs can be installed. Assuming that anaconda is available, all the required programs can be installed using the following:

```
conda create -n flye -c bioconda -c conda-forge -c defaults flye
```
## Introduction
### This pipeline is compatabile with reads generated by Oxford Nanopore or PacBio.

#### Before we run FastQC, you should be on a compute node in an interactive session. Please run the following srun command if you are not on a compute node. An interactive session is very useful to test tools and workflows.
```
srun --pty -t 0-6:00 --mem 5G /bin/bash
```

Now, after you are on interactive mode, develop the scripts required for analysis and submit the jobs using"
```
sbatch "filename.sh"
```

And, the squeue command is used to pull up information about jobs in the queue, by default this command will list the job ID, partition, username, job status, number of nodes, and name of nodes for all jobs queued or running within SLURM.
```
squeue -u username
```
And if you want to end the interactive mode type *exit*

#### The files provided initially is ```raw_data.fq.gz```

#### Inorder to start the analysis, we need to activate the conda environment 
```conda activate flye```

# Assembly using FLYE
#### Create a script name ```flye.sh``` that has the command to perform assembly.
```
/home/guduru.g/miniconda3/envs/flye/bin/flye --nano-raw raw_data.fq -o out_flye -g 2.4k -t 10 -i 10
```
You can always do a trial test with iterations part (-i option) until you get a best genome assembly.

## outputs
```00-assembly```, ```20-repeat```, ```30-contigger```, ```10-consensus```, ```40-polishing```
```params.json```
```assembly_graph.gv```
```assembly_graph.gfa```
```assembly.fasta```
```assembly_info.txt```

### It also provides assembly stats
![image](https://github.com/Gayathri-Guduru/Denovo_assembly_flye/assets/98939664/4f2b7967-ae62-49a1-8b3f-7c9974e058fa)


#### Next step is mapping the reference genome to the assembly fasta file.
#### Download the fasta sequence from https://www.addgene.org/50005/sequences/ which is the reference fasta file.

# Extra Steps:
###(if you are not provided with a reference fasta sequence file)
#### Download the genbank sequence which is in the format .gbk, from https://www.addgene.org/50005/sequences/ and convert it to fasta format.
#### create a ```genbank_to_fasta.py``` python script to convert genbank sequence to fasta format
```
import warnings
from Bio import BiopythonParserWarning
warnings.simplefilter('ignore', BiopythonParserWarning)
from Bio import SeqIO                                                                                                                                                                                             # Replace this with the path to your GenBank file
genbank_file_path = "/home/guduru.g/flye/addgene-plasmid-50005-sequence-222046.gbk"                                                                                                                               # Replace this with the desired output path for the FASTA file
fasta_file_path = "/home/guduru.g/flye/reference_pUC19.fasta"
# Converting GenBank file to FASTA format
with open(genbank_file_path) as input_handle, open(fasta_file_path, "w") as output_handle:
sequences = SeqIO.parse(input_handle, "genbank")
count = SeqIO.write(sequences, output_handle, "fasta")                                                                                                                                                            print(f"Converted {count} records to FASTA format")
```
#### This provides ```reference_pUC19.fasta``` reference file that is used to map with assembly fasta file.

#### Now, for mapping we use ```MUMmer```
```/home/guduru.g/miniconda3/envs/bowtie2/bin/nucmer --prefix=out reference.fasta out_flye/assembly.fasta```

## Generate a Readable Report:
#### To view the alignment, you can use the show-coords utility:
```/home/guduru.g/miniconda3/envs/bowtie2/bin/show-coords -rcl out.delta > alignment_report.txt```

#### This creates a file named ```alignment_report.txt``` with a readable alignment report.

#### Perform ```blastN``` online to compare the mapping results got from MUMmer.
![image](https://github.com/Gayathri-Guduru/Denovo_assembly_flye/assets/98939664/81faf3b5-5236-4b15-a453-620602504e56)

#### The image displays BLASTn results showing a perfect match (100% identity, 0% gaps) between the query sequence and the cloning vector pUC19c (GenBank accession L09137.2), suggesting the query(our assembly) is identical to this well-characterized vector sequence.

#### Futher Circos is performed for visualizing and to show similarly of your query vs the reference sequence sequences at region wise.
![proksee-7997x4364](https://github.com/Gayathri-Guduru/Denovo_assembly_flye/assets/98939664/d61cb06a-d953-45a7-b113-8db330098425)

#### The image shows a circular map of the pUC19 plasmid which is 2,686 base pairs in size. The map features several concentric circles: the outermost circle represents the reference pUC19 sequence, the next one in blue indicates a de novo assembly for comparison. Inside, there are two tracks for GC content (black) and GC skew, with green representing positive skew and purple representing negative skew. The GC skew tracks show the asymmetry in the distribution of guanine and cytosine nucleotides. This kind of map is typically used in genomics to visualize and compare genomic features.

